src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs

```diff
@@ -45,32 +46,23 @@ private async Task<PingReply> SendPingAsyncCore(IPAddress address, byte[] buffer
             return reply;
         }
 
-        private SocketConfig GetSocketConfig(IPAddress address, byte[] buffer, int timeout, PingOptions options)
+        private SocketConfig GetSocketConfig(IPAddress address, byte[] buffer, int timeout, PingOptions? options)
         {
-            SocketConfig config = new SocketConfig();
-            config.EndPoint = new IPEndPoint(address, 0);
-            config.Timeout = timeout;
-            config.Options = options;
-
-            config.IsIpv4 = address.AddressFamily == AddressFamily.InterNetwork;
-            config.ProtocolType = config.IsIpv4 ? ProtocolType.Icmp : ProtocolType.IcmpV6;
-
             // Use a random value as the identifier. This doesn't need to be perfectly random
             // or very unpredictable, rather just good enough to avoid unexpected conflicts.
-            Random rand = t_idGenerator ?? (t_idGenerator = new Random());
-            config.Identifier = (ushort)rand.Next((int)ushort.MaxValue + 1);
+            Random rand = t_idGenerator ??= new Random();
+            ushort id = (ushort)rand.Next(ushort.MaxValue + 1);
 
-            IcmpHeader header = new IcmpHeader()
-            {
-                Type = config.IsIpv4 ? (byte)IcmpV4MessageType.EchoRequest : (byte)IcmpV6MessageType.EchoRequest,
-                Code = 0,
-                HeaderChecksum = 0,
-                Identifier = config.Identifier,
-                SequenceNumber = 0,
-            };
-
-            config.SendBuffer = CreateSendMessageBuffer(header, buffer);
-            return config;
+            bool ipv4 = address.AddressFamily == AddressFamily.InterNetwork;
+
+            return new SocketConfig(
+                new IPEndPoint(address, 0), timeout, options,
```

Comment:
Out of curiosity, what necessitated a refactoring here?

Reply:
The fields were all being assigned post-construction, which meant we would either need to make the fields nullable and ! on all subsequent usage, or we'd need to make them non-nullable and hope that the call site knew to set them.  By forcing everything to be passed to the constructor, we could make the fields appropriately non-nullable without needing to rely on the caller to know to set the fields.